synapse/logging/opentracing.py

```diff
@@ -807,6 +807,13 @@ def err_back(result):
                         result.addCallbacks(call_back, err_back)
 
                     else:
+                        if inspect.iscoroutine(result):
+                            logger.error(
+                                "@trace did not wrap %s correctly! "
+                                "The function is not async but returns a coroutine.",
+                                func.__qualname__,
+                            )
```

Comment:
Should this be an error or warning?

I chose error because I'd like it to show up in sentry. On the other hand it doesn't impact normal operation of the homserver.

Reply:
Should this raise an error since it is a programming error that shouldn't occur?

Reply:
Maybe an error would be more appropriate then. (Note that we do seem to handle coroutines properly above, so this would only happen if we have a function which returns an awaitable, but is not a coroutine. We do have some of those though...)

Reply:
Alright, I'll change it to an assert then.

Reply:
Sorry, I wasn't clear. I meant "maybe logging an error is more appropriate". As in -- don't change anything.

I don't feel strongly either way.

Also, shouldn't we be checking via `inspect.isawaitable`, not `iscoroutine` here?